fix(plausible): migrate to new common chart and upgrade to v2

schnerring opened the pull request:
**Description**

* cherry-picked changes from my PR that were never merged into from `port/incubator` into `master`: https://github.com/truecharts/charts/pull/7810 
* Added config options for Plausible v2.0.0
* Migrated to new common chart

TODO:

- [x] upgrade questions.yaml to new common chart
- [x] migrate common.yaml / _env.tpl to new common chart format

⚒️ Fixes  # <!--(issue)-->

**⚙️ Type of change**

- [ ] ⚙️ Feature/App addition
- [X] 🪛 Bugfix
- [X] ⚠️ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] 🔃 Refactor of current code

**🧪 How Has This Been Tested?**
<!--
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
-->

**📃 Notes:**
<!-- Please enter any other relevant information here -->

**✔️ Checklist:**

- [X] ⚖️ My code follows the style guidelines of this project
- [X] 👀 I have performed a self-review of my own code
- [X] #️⃣ I have commented my code, particularly in hard-to-understand areas
- [ ] 📄 I have made corresponding changes to the documentation
- [ ] ⚠️ My changes generate no new warnings
- [ ] 🧪 I have added tests to this description that prove my fix is effective or that my feature works
- [X] ⬆️ I increased versions for any altered app according to semantic versioning

**➕ App addition**

If this PR is an app addition please make sure you have done the following.

- [ ] 🪞 I have opened a PR on [truecharts/containers](https://github.com/truecharts/containers) adding the container to TrueCharts mirror repo.
- [ ] 🖼️ I have added an icon in the Chart's root directory called `icon.png`

---

_Please don't blindly check all the boxes. Read them and only check those that apply.
Those checkboxes are there for the reviewer to see what is this all about and
the status of this PR with a quick glance._

schnerring committed:
fix(plausible): Remove deprecated config options

schnerring committed:
fix(plausible): migrate Chart.yaml to new common chart

schnerring committed:
fix(plausible): use image from tccr.io

schnerring committed:
fix(plausible): migrate values.yaml to new common chart and use CNPG

schnerring committed:
fix(plausible): add v2 configuration options

schnerring commented on charts/incubator/plausible/questions.yaml:
```diff
@@ -134,13 +126,78 @@ questions:
+          schema:
+            show_if: [["MAXMIND_LICENSE_KEY", "!=", ""]]
+            type: string
+            required: true
```
Honestly not sure if this works together with `show_if`. Should I just remove `required: true`?

stavros-k replied:
It does work! It will only be required if the licence key is not empty!

schnerring commented on charts/incubator/plausible/values.yaml:
```diff
@@ -31,6 +29,18 @@ plausible:
+  # todo
+  # GEOLITE2_COUNTRY_DB: ""
+  # GEONAMES_SOURCE_FILE: ""
 
```
Plausible v2 integrates automatic MaxMind GeoLite2 updates (optional). However, I don't know where these updates are downloaded to and how big these updates are. If this is downloaded to ephemeral storage, after each container restart, these updates will be downloaded again. To optimize this, we would have to do some digging in the related PR and configure the storage accordingly: https://github.com/plausible/analytics/pull/2362

schnerring committed:
fix(plausible): fix linting errors

schnerring committed:
fix(plausible): migrate template files to new common format

schnerring committed:
fix(plausible): fix CLICKHOUSE_DATABASE_URL

schnerring committed:
fix(plausible): remove invalid service protocol

schnerring commented:
Any idea what to do about this linting error?

```text
Error: execution error at (plausible/templates/common.yaml:2:4): Operator [cloudnative-pg] has to be installed first
```

stavros-k commented:
> Any idea what to do about this linting error?
> 
> ```
> Error: execution error at (plausible/templates/common.yaml:2:4): Operator [cloudnative-pg] has to be installed first
> ```

Commented out the helm template from the linting stage, as it would require operators to be installed there as well.
Should be passing now

stavros-k commented on charts/incubator/plausible/values.yaml:
```diff
@@ -1,28 +1,26 @@
+  DISABLE_REGISTRATION: "true"
+  LOG_FAILED_LOGIN_ATTEMPTS: false
+  # todo(this is fine for SCALE, but not for other multi-node platforms)
+  # RELEASE_DISTRIBUTION: none
```
You can leave this commented here, but not expose in the UI.
If someone is "advanced" enough, can add it on scale with "Additional envs"

> I have no idea what this does :D

schnerring replied:
From the Plausible docs:

> Alternatively, if you are running on a single machine and don't need any of Erlang's multi-node distribution features, you turn them off entirely by setting environment variable `RELEASE_DISTRIBUTION` to `none`

See also: https://hexdocs.pm/mix/Mix.Tasks.Release.html#module-environment-variables

This would be a optimization to set this to `none` for undistributed deployments (i.e. SCALE). But it's probably best to keep it simple and just leave these advanced configs for advanced users.

stavros-k commented on charts/incubator/plausible/templates/_secret.tpl:
```diff
@@ -0,0 +1,28 @@
+  {{- else }}
+  {{- /* The plain value of SECRET_KEY_BASE is also base64 encoded */}}
+  SECRET_KEY_BASE: {{ randAlphaNum 65 | b64enc }}
+  {{- end }}
```
```suggestion
{{- /* The plain value of SECRET_KEY_BASE is also base64 encoded */}}
{{- $key := randAlphaNum 65 | b64enc -}}
{{- with (lookup "v1" "Secret" .Release.Namespace "plausible-secret") -}}
  {{- $key = index .data "SECRET_KEY_BASE" | b64dec -}}
{{- end }}
enabled: true
data:
  SECRET_KEY_BASE: {{ $key }}
```

Personally I like this "better" but I'm fine either way.

Only "nit", it's `data` not `stringData` Even tho content is added as stringData..
(It should have been something like `content` but it's too late now)

Just to be sure, the length is correctly `65` right? (I usually see those being multiples of 2).

schnerring replied:
I found both versions throughout the code base, so I stuck with the one I liked more.

> Just to be sure, the length is correctly 65 right? (I usually see those being multiples of 2).

Not sure how I came up with a length `65` last time. 😅 Plausible recommends using the default [`mix phx.gen.secret`](https://hexdocs.pm/phoenix/Mix.Tasks.Phx.Gen.Secret.html#content) of the Phoenix Framework or `openssl rand -hex 64` to generate the secret.

---

Just because I find it a fun exercise:

`openssl rand -hex 64` generates a 128 character hex string, thus a secret with _512 bits_ of entropy.

[To find the length, _L_, needed to achieve a desired strength _H_, with a password drawn randomly from a set of _N_ symbols, one computes](https://en.wikipedia.org/wiki/Password_strength#Entropy_as_a_measure_of_password_strength):

$$ L = \lceil \dfrac{H}{log_{2} N} \rceil $$

To get a secret with strength _H = 512_ with Helm's `randAlphaNum` function where N = 62:

$$ L = \lceil \dfrac{512}{log_{2} 62} \rceil = 86 $$

So let's go with `86` isntead!

So the multiples of 2 really only make sense for hex strings:

![image](https://github.com/truecharts/charts/assets/3743342/0579aecb-10f6-42bc-b352-71efff993036)

stavros-k commented on charts/incubator/plausible/values.yaml:
```diff
@@ -39,72 +50,78 @@ plausible:
+                name: plausible-config
+            - secretRef:
+                name: plausible-secret
+          command: [sh]
```
fyi, if its a single item, you can leave it as string.
But if you leave it as list shouldn't it be `["sh"]` ?
(or yaml list)

stavros-k commented:
For catalog test failing, you can "generate" questions and see for errors
From the root of the project
```
./tools/gen-cat.sh charts/incubator/plausible
```
Output will go to a new dir on the root of the project called `catalog`

stavros-k commented on charts/incubator/plausible/values.yaml:
```diff
@@ -1,28 +1,26 @@
+  repository: tccr.io/truecharts/plausible
   pullPolicy: IfNotPresent
-  tag: v1.5.1@sha256:3fbdcfac56ffa79e778e9e4bc1ca8c4dca0d7deb330ed5694f59eaa1125cc0a8
+  tag: v2.0.0@sha256:5bf2057948e1d5d12ef61f446c4945a336d3bcdf9b96293a1d6fc275f3c56eb5
```
```suggestion
  tag: v2.0.0@sha256:1ef59e863cfdcddc308d4ec743091bc8eda4f62e16b269cf4d778c89b8c02a68
```

stavros-k committed:
Update charts/incubator/plausible/values.yaml

Signed-off-by: Stavros Kois <47820033+stavros-k@users.noreply.github.com>

schnerring committed:
fix(plausible): implement code review suggestions

schnerring commented:
> ```
> ./tools/gen-cat.sh charts/incubator/plausible
> ```

That command returns the following error when running it:

```text
Error: no cached repository for helm-manager-abaa37bfaf705b7726f614114b77dd23196b42f5087901e57467be7fa3fead01 found. (try 'helm repo update'): open /home/***/.cache/helm/repository/helm-manager-abaa37bfaf705b7726f614114b77dd23196b42f5087901e57467be7fa3fead01-index.yaml: no such file or directory
```

Any idea what I'm missing?

stavros-k commented:
> > ```
> > ./tools/gen-cat.sh charts/incubator/plausible
> > ```
> 
> That command returns the following error when running it:
> 
> ```
> Error: no cached repository for helm-manager-abaa37bfaf705b7726f614114b77dd23196b42f5087901e57467be7fa3fead01 found. (try 'helm repo update'): open /home/***/.cache/helm/repository/helm-manager-abaa37bfaf705b7726f614114b77dd23196b42f5087901e57467be7fa3fead01-index.yaml: no such file or directory
> ```
> 
> Any idea what I'm missing?

Mine seem to work, maybe you need to add the repositories

```shell
❯ ./tools/gen-cat.sh charts/incubator/plausible 
No owner configured, defaulting to truecharts
No repo configured, defaulting to charts
env #CR_TOKEN not found, defaulting to production=false
"truecharts" already exists with the same configuration, skipping
"truecharts-library" already exists with the same configuration, skipping
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "truecharts-library" chart repository
...Successfully got an update from the "truecharts-deps" chart repository
...Successfully got an update from the "truecharts" chart repository
...Successfully got an update from the "bitnami" chart repository
...Successfully got an update from the "prometheus" chart repository
Update Complete. ⎈Happy Helming!⎈
Start processing charts/incubator/plausible ...
Saving 3 charts
Downloading common from repo https://library-charts.truecharts.org
Downloading postgresql from repo https://deps.truecharts.org/
Downloading clickhouse from repo https://deps.truecharts.org/
Deleting outdated charts
Cleaning SCALE catalog for Chart: plausible
Copying Chart to Catalog: plausible
Applying SCALE patches for Chart: plausible
Making copy of charts/incubator/plausible/questions.yaml to catalog/incubator/plausible/4.0.25/questions.yaml
Including standardised questions.yaml includes for: plausible
Removing old versions for: plausible
/home/skois/projects/truecharts/apps
Done processing charts/incubator/plausible ...
Starting post-processing
```

schnerring committed:
Add deps.truecharts.org helm repo to tools/gen-cat.sh

schnerring committed:
fix(plausible): migrate questions.yaml to new common chart layout

schnerring commented:
> Mine seem to work, maybe you need to add the repositories

`tools/get-cat.sh` was missing the `truecharts-deps` repo, I added it to the `prep_helm` function of the script.

Regarding questions.yaml, I didn't really know what I was doing / looking for, so I just did a side-by-side comparison with the up-to-date `paperless-ngx` questions.yaml and changed the `plausible` questions.yaml accordingly.

Now the "install charts" CI step fails. It looks like ClickHouse is the issue. Here's the event log of the `plausible-luh9hueop1-clickhouse-5bd7db77f8-qnfql` pod:

```text
Events:
  Type     Reason     Age                    From               Message
  ----     ------     ----                   ----               -------
  Normal   Scheduled  4m4s                   default-scheduler  Successfully assigned plausible-luh9hueop1/plausible-luh9hueop1-clickhouse-5bd7db77f8-qnfql to k3d-k3s-default-server-0
  Normal   Pulling    4m3s                   kubelet            Pulling image "tccr.io/truecharts/clickhouse:v23.7.1.2470@sha256:fbec51f0c6b09733fe4601bba2584dc750f0a8a5077702b86163744bd6f4573f"
  Normal   Pulled     3m46s                  kubelet            Successfully pulled image "tccr.io/truecharts/clickhouse:v23.7.1.2470@sha256:fbec51f0c6b09733fe4601bba2584dc750f0a8a5077702b86163744bd6f4573f" in 16.389378822s
  Normal   Pulled     102s (x11 over 3m46s)  kubelet            Container image "tccr.io/truecharts/clickhouse:v23.7.1.2470@sha256:fbec51f0c6b09733fe4601bba2584dc750f0a8a5077702b86163744bd6f4573f" already present on machine
  Warning  Failed     102s (x12 over 3m46s)  kubelet            Error: secret "clickhousecreds" not found
```

I guess this is due the common chart refactorings to the ClickHouse chart / library-chart? Gotta do some more debugging..........

schnerring committed:
fix(plausible): don't use existingSecret for ClickHouse

schnerring committed:
fix(plausible): upgrade ClickHouse chart dependency

schnerring commented:
Things are looking good! However, CNPG fails to start up in the _Install charts Part 1_ tests. Any idea why? I don't see anything in the logs.

```text
 Events:
  Type     Reason     Age                    From               Message
  ----     ------     ----                   ----               -------
  Normal   Scheduled  3m36s                  default-scheduler  Successfully assigned plausible-vu2tnxzq9l/plausible-vu2tnxzq9l-cnpg-main-1 to k3d-k3s-default-server-0
  Normal   Pulled     3m37s                  kubelet            Container image "tccr.io/truecharts/cloudnative-pg:v1.20.2@sha256:07121d674c1de587044746f92cc3bccb4576f2fc125f6f2f5bcaa582dad3e3ad" already present on machine
  Normal   Created    3m37s                  kubelet            Created container bootstrap-controller
  Normal   Started    3m36s                  kubelet            Started container bootstrap-controller
  Normal   Pulled     3m36s                  kubelet            Container image "ghcr.io/cloudnative-pg/postgresql:15.3" already present on machine
  Normal   Created    3m36s                  kubelet            Created container postgres
  Normal   Started    3m36s                  kubelet            Started container postgres
  Warning  Unhealthy  3m33s (x3 over 3m35s)  kubelet            Readiness probe failed: HTTP probe failed with statuscode: 500
```

stavros-k commented:
> Things are looking good! However, CNPG fails to start up in the _Install charts Part 1_ tests. Any idea why? I don't see anything in the logs.
> 
> ```
>  Events:
>   Type     Reason     Age                    From               Message
>   ----     ------     ----                   ----               -------
>   Normal   Scheduled  3m36s                  default-scheduler  Successfully assigned plausible-vu2tnxzq9l/plausible-vu2tnxzq9l-cnpg-main-1 to k3d-k3s-default-server-0
>   Normal   Pulled     3m37s                  kubelet            Container image "tccr.io/truecharts/cloudnative-pg:v1.20.2@sha256:07121d674c1de587044746f92cc3bccb4576f2fc125f6f2f5bcaa582dad3e3ad" already present on machine
>   Normal   Created    3m37s                  kubelet            Created container bootstrap-controller
>   Normal   Started    3m36s                  kubelet            Started container bootstrap-controller
>   Normal   Pulled     3m36s                  kubelet            Container image "ghcr.io/cloudnative-pg/postgresql:15.3" already present on machine
>   Normal   Created    3m36s                  kubelet            Created container postgres
>   Normal   Started    3m36s                  kubelet            Started container postgres
>   Warning  Unhealthy  3m33s (x3 over 3m35s)  kubelet            Readiness probe failed: HTTP probe failed with statuscode: 500
> ```

I'm not sure that its stays Unhealthy, its just the last event logged. If it stayed unhealthy it would be somelike like x20-x30 over the time.

What I've seen is this

```shell
------------------------------------------------------------------------------------------------------------------------
==> Logs of init container plausible-vu2tnxzq9l-c6665c7b6-8rlr8
------------------------------------------------------------------------------------------------------------------------
Loading plausible..
Starting dependencies..
Starting repos..
create Plausible.Repo database if it doesn't exist
create Plausible.IngestRepo database if it doesn't exist
Creation of Db successful!
Loading plausible..
Starting dependencies..
Starting repos..
Running migrations for Elixir.Plausible.Repo
Running migrations for Elixir.Plausible.IngestRepo
Migrations successful!
------------------------------------------------------------------------------------------------------------------------
<== Logs of init container plausible-vu2tnxzq9l-c6665c7b6-8rlr8
------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------
==> Logs of init container plausible-vu2tnxzq9l-c6665c7b6-8rlr8
------------------------------------------------------------------------------------------------------------------------
/bin/sh: /bin/bash: not found
------------------------------------------------------------------------------------------------------------------------
<== Logs of init container plausible-vu2tnxzq9l-c6665c7b6-8rlr8
------------------------------------------------------------------------------------------------------------------------
```

Is there any chance the migration entrypoint.sh trying to use bash?

stavros-k commented:
@schnerring Found the issue
https://github.com/truecharts/library-charts/commit/48b8e123edb431d114e402f51706d1a5b3d4380f

stavros-k commented on charts/incubator/plausible/Chart.yaml:
```diff
@@ -1,17 +1,13 @@
-    name: postgresql
-    repository: https://deps.truecharts.org/
-    version: 11.0.31
+    version: 14.0.1
```
```suggestion
    version: 14.0.2
```

stavros-k committed:
Update charts/incubator/plausible/Chart.yaml

Signed-off-by: Stavros Kois <47820033+stavros-k@users.noreply.github.com>

stavros-k committed:
Merge branch 'master' into upgrade-plausible

schnerring commented:
> @schnerring Found the issue [truecharts/library-charts@48b8e12](https://github.com/truecharts/library-charts/commit/48b8e123edb431d114e402f51706d1a5b3d4380f)

Nice find, thanks for the help!

truecharts-admin commented:
This PR is locked to prevent necro-posting on closed PRs. Please create a issue or contact staff on discord if you want to further discuss this
